compiler/sem/lookups.nim

```diff
@@ -54,50 +54,158 @@ proc ensureNoMissingOrUnusedSymbols(c: PContext; scope: PScope)
 
+  proc accQuotedConstrIdentNode(c: PContext, n: PNode): PNode =
+    ## construct and identifier the `nkAccQuoted` node `n`, returning an
+    ## `nkIdent` node, or `nkError` upon failure.
+    const limit = 20 ## limit how much evaluation we do here
+    assert n.kind == nkAccQuoted
+    var
+      stack: seq[(int, PNode)] = @[(0, n)]
+        ## tracks the depth first node traversal (child index and node) of
+        ## `nkAccQuoted` nodes, the index is mutated to track which quoted
+        ## child needs to be processed next
+      wellFormedAst = true
+        ## tracks if `n` is well formed AST even if there are no `nkError`
+      hasError = false
+        ## `n` already has errors (`nkError` nodes)
+      counter = 0
+        ## how many iterations we've done, used to avoid lengthy/endless eval
+      flattened: seq[PNode]
+        ## collection point for a flat `nkAccQuote` tree, for ident string gen
+
+    while stack.len > 0 and counter < limit:
+      let
+        stackPos = stack.len - 1
+        quote = stack[stackPos][1]
+      var idx = stack[stackPos][0]
+
+      # process until the next `nkAccQuoted` that requires descent
+      while idx < quote.len and
+            (quote[idx].kind != nkAccQuoted or quote[idx].len == 0):
+        inc counter
+        let q = quote[idx]
+        case q.kind
+        of atomicIdentKinds, nkIntLiterals, nkStrLiterals:
+          flattened.add q
+        of nkSymChoices:
+          flattened.add:
+            if q.len > 0 and q[0].kind == nkSym:
+              q[0]
+            else:
+              q   # treat `nkSymChoices` as an error signal below
+        of nkAccQuoted:
+          assert q.len == 0
+          flattened.add q
+        of nkError:
+          hasError = true
+          flattened.add q
+        of allNodeKinds - nkIdentKinds - renderableKinds - {nkError}:
+          wellFormedAst = false
+          flattened.add q
+        inc idx
+
+      if idx < quote.len:
+        assert quote[idx].kind == nkAccQuoted and quote[idx].len > 0
+        stack.add (0, quote[idx])
+        inc counter
+        inc idx
+        stack[stackPos][0] = idx # update to latest index
+      else:
+        discard stack.pop() # we're done processing pop and move on
+
+    # now we have a very flattened/normalized `nkAccQuoted` ast expression,
+    # time to try constructing an identifier.
+    result =
+      if hasError:
+        c.config.wrapError(n)
+      elif counter >= limit:
+        c.config.newError(n, PAstDiag(kind: adSemExpectedIdentifierQuoteLimit))
+      elif wellFormedAst:
+        var
+          id = ""
+          badNode: PNode = nil
+        for s in flattened.items:
+          case s.kind
+          of nkIdent:                     id.add(s.ident.s)
+          of nkSym:                       id.add(s.sym.name.s)
+          of nkStrLiterals:               id.add(s.strVal)
+          of nkCharLit:                   id.add(char(s.intVal))
```

Comment:
Nothing that has to happen now, but I think in the future, we should probably disallow characters outside of the 32..127 range.

Reply:
Yup, i was able to put new lines in identifier and all sorts of fun stuff.